File: baseunits/modules/JapScan.pas
unit JapScan;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread;

implementation

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo.FHTTP.GET(Module.RootURL + '/mangas/') then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        XPathHREFAll('//*[@id="liste_mangas"]/div[@class="row"]/div[@class="cell"]/a', ALinks, ANames);
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  v: IXQValue;
  s: String;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo, MangaInfo.FHTTP do
  begin
    url := MaybeFillHost(Module.RootURL, AURL);
    if GET(url) then begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do
        try
          title := XPathString('//title');
          if Pos(' | Japscan.Com', title) <> 0 then
            title := GetString(title, 'Lecture En Ligne Des Chapitres ', ' | Japscan.Com');
          authors := XPathString('//div[@class="table"]/div[@class="row"]/div[1]');
          genres := XPathString('//div[@class="table"]/div[@class="row"]/div[4]');
          summary := XPathString('//div[@id="synopsis"]/string-join(text(),codepoints-to-string(10))');
          status := MangaInfoStatusIfPos(XPathString('//div[@class="table"]/div[@class="row"]/div[6]'), 'En Cours', 'Terminé');
          for v in XPath('//*[@id="liste_chapitres"]/ul/li/a') do
          begin
            chapterLinks.Add(v.toNode.getAttribute('href'));
            s := v.toString;
            if Pos('[email protected]', s) <> 0 then
              s := title + ' ' + XPathString('text()[2]', v)
            else
            if Pos('Scan ', s) = 1 then
              Delete(s, 1, 5);
            chapterName.Add(s);
          end;
          InvertStrings([chapterLinks, chapterName]);
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  v: IXQValue;
  dataimg, imgurl: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.Task.Container, DownloadThread.FHTTP do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(MaybeFillHost(Module.RootURL, AURL)) then
    begin
      Result := True;
      XPathStringAll('//select[@id="pages"]/option/@data-img', Document, PageLinks);
    end;
  end;
end;

function DownloadImage(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
begin
  Result := False;
  with DownloadThread, DownloadThread.FHTTP, DownloadThread.Task.Container do
  begin
    if GET(AppendURLDelim(FillHost(Module.RootURL, ChapterLinks[CurrentDownloadChapterPtr])) + IncStr(WorkId) + '.html') then
      Result := GET(XPathString('//img[@id="image"]/@src', Document));
  end;
end;

procedure RegisterModule;
begin
  with AddModule do
  begin
    Website := 'Japscan';
    RootURL := 'http://www.japscan.com';
    Category := 'French';
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
    OnDownloadImage := @DownloadImage;
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/JapScan.lua
local ALPHA_LIST_UP = '#ABCDEFGHIJKLMNOPQRSTUVWXYZ'

function getinfo()
  mangainfo.url=MaybeFillHost(module.RootURL, url)
  if http.get(mangainfo.url) then
    local x=TXQuery.Create(http.document)
    if mangainfo.title == '' then
      mangainfo.title = x.xpathstring('//h1')
      mangainfo.title = string.gsub(mangainfo.title, '^Manga ', '')
      mangainfo.title = string.gsub(mangainfo.title, '^Manhua ', '')
      mangainfo.title = string.gsub(mangainfo.title, '^Manhwa ', '')
      mangainfo.title = string.gsub(mangainfo.title, ' VF$', '')
    end
    mangainfo.coverlink=MaybeFillHost(module.RootURL, x.xpathstring('//div[@id="main"]//img/@src'))
    mangainfo.status = MangaInfoStatusIfPos(x.xpathstringall('//div[@id="main"]//p[contains(span, "Statut")]/text()', ''), 'En Cours', 'Termine')
    mangainfo.authors=x.xpathstringall('//div[@id="main"]//p[contains(span, "Auteur")]/text()', '')
    mangainfo.artists=x.xpathstringall('//div[@id="main"]//p[contains(span, "Artiste")]/text()', '')
    mangainfo.genres=x.xpathstringall('//div[@id="main"]//p[contains(span, "Type(s)")]/text()', '')
    mangainfo.summary=x.xpathstring('//div[@id="main"]//div[contains(text(), "Synopsis")]/following-sibling::*')
    x.xpathhrefall('css("div#chapters_list div.chapters_list > a")', mangainfo.chapterlinks, mangainfo.chapternames)
    InvertStrings(mangainfo.chapterlinks, mangainfo.chapternames)
    return no_error
  else
    return net_problem
  end
end

function getpagenumber()
  task.pagelinks.clear()
  if http.get(MaybeFillHost(module.rooturl, url)) then
    local x=TXQuery.Create(http.Document)
    task.pagenumber = x.xpathcount('//select[@id="pages"]/option/@value')
  else
    return false
  end
  return true
end

function getimageurl()
  local s = AppendURLDelim(url)..(workid+1)..'.html'
  if http.get(MaybeFillHost(module.rooturl,s)) then
    task.pagelinks[workid]=TXQuery.create(http.document).xpathstring('//div[@id="image"]/@data-src')
    return true
  end
  return false
end

function getnameandlink()
  local s = '0-9'
  if module.CurrentDirectoryIndex ~= 0 then
    s = ALPHA_LIST_UP:sub(module.CurrentDirectoryIndex+1,module.CurrentDirectoryIndex+1)
  end
  if http.get(MaybeFillHost(module.rooturl, '/mangas/' .. s .. '/' .. IncStr(url))) then
    local x = TXQuery.Create(http.Document)
    x.XPathHREFAll('//div[@id="main"]//p/a[contains(@href, "/manga/")]', links, names)
    local page = tonumber(x.XPathString('//ul[contains(@class, "pagination")]/li[last()]/a'))
    if page == nil then
      page = 1
    end
    updatelist.CurrentDirectoryPageNumber = page
    return no_error
  else
    return net_problem
  end
end

function Init()
  local m=NewModule()
  m.category='French'
  m.website='Japscan'
  m.rooturl='http://www.japscan.co'
  m.lastupdated='April 6, 2018'
  m.ongetinfo='getinfo'
  m.ongetpagenumber='getpagenumber'
  m.ongetnameandlink='getnameandlink'
  m.ongetimageurl='getimageurl'
  m.totaldirectory=ALPHA_LIST_UP:len()
end
